include/cetl/any.hpp

```diff
@@ -0,0 +1,345 @@
+/// @file
+/// Defines the C++17 `std::any` type and several related entities.
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef CETL_ANY_HPP_INCLUDED
+#define CETL_ANY_HPP_INCLUDED
+
+#include "pf17/cetlpf.hpp"
+#include "pf17/utility.hpp"
+#include "pf17/attribute.hpp"
+
+#include <algorithm>
+#include <type_traits>
+
+namespace cetl
+{
+
+// Forward declarations
+template <std::size_t Footprint, bool Copyable, bool Movable>
+class any;
+
+namespace detail
+{
+
+// Move policy.
+template <std::size_t Footprint, bool Copyable, bool Movable>
+struct base_move;
+
+// Copy policy.
+template <std::size_t Footprint, bool Copyable>
+struct base_copy;
+
+template <std::size_t Footprint, std::size_t Align = sizeof(std::max_align_t)>
+struct base_storage
+{
+    // We need to align the buffer to the given value (maximum alignment by default).
+    // Also, we need to ensure that the buffer is at least 1 byte long.
+    alignas(Align) char buffer_[std::max(Footprint, 1UL)];
+};
+
+// Movable case.
+template <std::size_t Footprint, bool Copyable>
+struct base_move<Footprint, Copyable, true> : base_copy<Footprint, Copyable>
+{
+public:
+    constexpr base_move()                      = default;
+    base_move(base_move&&) noexcept            = default;
+    base_move& operator=(base_move&&) noexcept = default;
+};
+
+// Non-movable case.
+template <std::size_t Footprint, bool Copyable>
+struct base_move<Footprint, Copyable, false> : base_copy<Footprint, Copyable>
+{
+public:
+    constexpr base_move()                      = default;
+    base_move(base_move&&) noexcept            = delete;
+    base_move& operator=(base_move&&) noexcept = delete;
+};
+
+// Copyable case.
+template <std::size_t Footprint>
+struct base_copy<Footprint, true> : base_storage<Footprint>
+{
+public:
+    constexpr base_copy()                  = default;
+    base_copy(const base_copy&)            = default;
+    base_copy& operator=(const base_copy&) = default;
+};
+
+// Non-copyable case.
+template <std::size_t Footprint>
+struct base_copy<Footprint, false> : base_storage<Footprint>
+{
+public:
+    constexpr base_copy()                  = default;
+    base_copy(const base_copy&)            = delete;
+    base_copy& operator=(const base_copy&) = delete;
+};
+
+enum class action
+{
+    Get,
+    Copy,
+    Move,
+    Destroy
+};
```

Comment:
This code is fine but really cool kids from the block use algebraic types for this these days, because it allows one to express more concepts through the type system:

```c++
struct action_get{};
struct action_copy{ any* other; };  // notice the `other`!
struct action_move{ any* other; };  // it is introduced only when needed
struct action_destroy{};
using action = variant<action_get, action_copy, action_move, action_destroy>;
```

Then you use `visit` to handle `action` in a type-safe way without the need to introduce unused parameters.

Just saying.

Reply:
I'll think about it. Also, maybe when I start working on Copyable/Movable constraints then it's possible that enum and switch won't be applicable anymore...